types/result.go

```diff
@@ -1,121 +1,170 @@
-	return r.Code == CodeTypeOK
+type ResultSetOption struct {
+	Code uint32 `json:"code,omitempty"`
+	// bytes data = 2;
+	Log  string `json:"log,omitempty"`
+	Info string `json:"info,omitempty"`
 }
 
-// IsErr returns true if Code is something other than OK.
-func (r ResponseQuery) IsErr() bool {
-	return r.Code != CodeTypeOK
+func FromResultSetOption(res ResultSetOption) ResponseSetOption {
+	return ResponseSetOption(res)
 }
 
-//---------------------------------------------------------------------------
-// override JSON marshalling so we dont emit defaults (ie. disable omitempty)
-// note we need Unmarshal functions too because protobuf had the bright idea
-// to marshal int64->string. cool. cool, cool, cool: https://developers.google.com/protocol-buffers/docs/proto3#json
+type ResultInitChain struct {
+	Validators []Validator `json:"validators"`
+}
 
-var (
-	jsonpbMarshaller = jsonpb.Marshaler{
-		EnumsAsInts:  true,
-		EmitDefaults: false,
+func FromResultInitChain(res ResultInitChain) ResponseInitChain {
+	vals := valsToPointers(res.Validators)
+	return ResponseInitChain{
+		Validators: vals,
 	}
-	jsonpbUnmarshaller = jsonpb.Unmarshaler{}
-)
+}
 
-func (r *ResponseSetOption) MarshalJSON() ([]byte, error) {
-	s, err := jsonpbMarshaller.MarshalToString(r)
-	return []byte(s), err
+type ResultQuery struct {
+	Code uint32 `json:"code,omitempty"`
+	// bytes data = 2; // use "value" instead.
+	Log    string `json:"log,omitempty"`
+	Info   string `json:"info,omitempty"`
+	Index  int64  `json:"index,omitempty"`
+	Key    []byte `json:"key,omitempty"`
+	Value  []byte `json:"value,omitempty"`
+	Proof  []byte `json:"proof,omitempty"`
+	Height int64  `json:"height,omitempty"`
 }
 
-func (r *ResponseSetOption) UnmarshalJSON(b []byte) error {
-	reader := bytes.NewBuffer(b)
-	return jsonpbUnmarshaller.Unmarshal(reader, r)
+func FromResultQuery(res ResultQuery) ResponseQuery {
+	return ResponseQuery(res)
 }
 
-func (r *ResponseCheckTx) MarshalJSON() ([]byte, error) {
-	s, err := jsonpbMarshaller.MarshalToString(r)
-	return []byte(s), err
+type ResultBeginBlock struct {
+	Tags []cmn.KVPair `json:"tags,omitempty"`
 }
 
-func (r *ResponseCheckTx) UnmarshalJSON(b []byte) error {
-	reader := bytes.NewBuffer(b)
-	return jsonpbUnmarshaller.Unmarshal(reader, r)
+func FromResultBeginBlock(res ResultBeginBlock) ResponseBeginBlock {
+	tags := tagsToPointers(res.Tags)
+	return ResponseBeginBlock{
+		Tags: tags,
+	}
 }
 
-func (r *ResponseDeliverTx) MarshalJSON() ([]byte, error) {
-	s, err := jsonpbMarshaller.MarshalToString(r)
-	return []byte(s), err
+type ResultCheckTx struct {
+	Code      uint32       `json:"code,omitempty"`
+	Data      []byte       `json:"data,omitempty"`
+	Log       string       `json:"log,omitempty"`
+	Info      string       `json:"info,omitempty"`
+	GasWanted int64        `json:"gas_wanted,omitempty"`
+	GasUsed   int64        `json:"gas_used,omitempty"`
+	Tags      []cmn.KVPair `json:"tags,omitempty"`
+	Fee       cmn.KI64Pair `json:"fee"`
```

Comment:
how do you choose whenever to add `omitempty` or not?

Reply:
it correlates with `(gogoproto.nullable)=false`, right?

Reply:
oh good point. I didn't choose. I just left it how it was. yeh i think it does correlate